### PR TITLE
feat(embodied): add DSRL support for Pi0

### PR DESCRIPTION
**Pull request by Elessar123**
### Description

Integrate [DSRL (Diffusion Steering via Reinforcement Learning)](https://github.com/nakamotoo/dsrl_pi0) into RLinf for Pi0. Based on the paper [Steering Your Diffusion Policy with Latent Space Reinforcement Learning](https://arxiv.org/abs/2506.15799) (CoRL 2025, Wagenmaker et al.).

DSRL steers a pre-trained Pi0 diffusion policy by training a lightweight SAC agent in the latent noise space: the SAC agent generates noise, which is fed into Pi0's diffusion denoiser to produce real actions for robotic control tasks. All DSRL code paths are gated behind `use_dsrl` flags (default: `False`), so existing behavior is unchanged when disabled.

**New files (4):**
- `rlinf/models/embodiment/modules/compact_encoders.py` — Lightweight CNN/MLP encoders and 10-Q-head ensemble for DSRL critic (~500K params total)
- `rlinf/models/embodiment/modules/gaussian_policy.py` — SquashedNormal distribution and GaussianPolicy for SAC actor
- `examples/embodiment/config/model/pi0_dsrl.yaml` — DSRL-enabled model config
- `examples/embodiment/config/libero_spatial_dsrl_openpi.yaml` — Full DSRL training config

**Modified files (6):**
- `rlinf/models/embodiment/openpi/openpi_action_model.py` — SAC forward dispatch (`ForwardType.SAC` / `SAC_Q`), DSRL rollout pipeline with noise injection into `sample_actions`, DSRL image/state preprocessing, and `freeze_vlm` extension
- `rlinf/workers/actor/fsdp_sac_policy_worker.py` — DSRL parameter grouping for FSDP, float32 EMA shadow buffer for bf16 models, bootstrap reward/discount gated by `use_dsrl`, `train=True` gated behind `use_dsrl`
- `rlinf/envs/libero/libero_env.py` — Configurable step penalty reward (`use_step_penalty`)
- `rlinf/config.py` — Fix Hydra `null`-as-string edge case
- `examples/embodiment/config/model/pi0.yaml` — DSRL defaults (disabled)
- `examples/embodiment/config/env/libero_spatial.yaml` — `use_step_penalty` default

**Bug fixes (benefit all SAC users):**
- `alpha_lr_scheduler` was using `self.optimizer` instead of `self.alpha_optimizer`
- `target_q_values` dtype misalignment: boolean ops with Python floats promote to float32, mismatching bf16 model outputs

### Motivation and Context

DSRL enables RL-based fine-tuning of pre-trained diffusion policies (Pi0) without modifying the frozen VLM backbone. By operating in the latent noise space rather than the action space, DSRL can steer Pi0 with a lightweight SAC agent (~500K trainable params) while preserving the pre-trained policy's generalization capabilities.

### How has this been tested?

- Model initialization and `freeze_vlm()` verified: DSRL components are correctly created and VLM parameters are frozen with non-zero trainable params remaining
- Config naming alignment between code and YAML validated
- Non-DSRL SAC (`maniskill_sac_mlp.yaml`) verified: no regression from gated `train=True`
- All files pass `ruff check --preview` and `ruff format`
- Full end-to-end DSRL training loop not yet validated (pending GPU cluster access)

### Additional information

- Reference: [dsrl_pi0](https://github.com/nakamotoo/dsrl_pi0) (JAX-based original). This PR is a PyTorch re-implementation integrated into RLinf's Ray/FSDP distributed training framework.
- 10 files changed, +1473 / -62 lines.

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/models/embodiment/openpi/openpi_action_model.py [hunk 1]
```diff
@@ -347,32 +421,97 @@ def predict_action_batch(
...
+                    else env_obs["wrist_images"]
+                )
+
+            result = {
+                "prev_logprobs": noise_logprob,  # SAC noise logprob
+                "forward_inputs": forward_inputs,
+            }
+
+            # Only include prev_values if it exists and is valid
+            if "prev_values" in outputs and outputs["prev_values"] is not None:
```
Comment (andylin-hao) [inline rlinf/models/embodiment/openpi/openpi_action_model.py]: The `generate_one_epoch` function in huggingface worker unconditionally reads result["prev_values"] (Line 272). If the condition triggers here, will the hugginggace worker fail to retrieve prev_values?

File: examples/embodiment/config/libero_spatial_dsrl_openpi.yaml [hunk 2]
```diff
@@ -0,0 +1,196 @@
+defaults:
+  - env/libero_spatial@env.train
+  - env/libero_spatial@env.eval
+  - model/pi0@actor.model
```
Comment (andylin-hao) [inline examples/embodiment/config/libero_spatial_dsrl_openpi.yaml:4]: Should it use the new `pi0_dsrl.yaml`?

Comment (Elessar123) [reply examples/embodiment/config/libero_spatial_dsrl_openpi.yaml:4]: Yes, fixed in the latest push — changed to `model/pi0_dsrl@actor.model`.

File: rlinf/models/embodiment/openpi/openpi_action_model.py (hunk 1 above)
Comment (Elessar123) [reply rlinf/models/embodiment/openpi/openpi_action_model.py]: Good catch. Fixed in the latest push — changed to `outputs.get("prev_values")` so that `result["prev_values"]` is always present.

In practice, DSRL uses SAC with `add_value_head: False`, so `sample_actions` returns `prev_values: None`. This is consistent with the non-DSRL branch (line 508) which also does `"prev_values": outputs["prev_values"]` unconditionally.

File: rlinf/models/embodiment/openpi/openpi_action_model.py [hunk 3]
```diff
@@ -337,7 +413,7 @@ def predict_action_batch(
         env_obs,
         mode: Literal["train", "eval"] = "train",
         compute_values=True,
-        **kwargs,
+        return_obs=True,
```
Comment (guozhen1997) [inline rlinf/models/embodiment/openpi/openpi_action_model.py]: Could you change this back to `**kwargs`? The current `return_obs=True` isn’t used anywhere.

Comment (Elessar123) [reply rlinf/models/embodiment/openpi/openpi_action_model.py]: Sure, I have fixed it.

**Review by guozhen1997 — APPROVED** [review]
